### PR TITLE
Add move event emitting staking and rewards info during epoch change

### PR DESCRIPTION
**Pull request by emmazzz**
This PR implements event emission during epoch changes.
One system level event `SystemEpochInfo` is emitted, and one validator level event `ValidatorEpochInfo` is emitted per validator.

### CODE REVIEW

**Comment by emmazzz** [comment]
addresses #6835

File: crates/sui-framework/sources/governance/sui_system.move [hunk 1]
```diff
@@ -460,6 +483,20 @@ module sui::sui_system {
         // Validator reports are only valid for the epoch.
         // TODO: or do we want to make it persistent and validators have to explicitly change their scores?
         self.validator_report_records = vec_map::empty();
+
+        event::emit(
+            SystemEpochInfo {
+                epoch: self.epoch,
+                reference_gas_price: self.reference_gas_price,
+                total_stake: delegation_stake + validator_stake,
+                storage_fund_inflows: storage_charge + (storage_fund_reinvestment_amount as u64),
```
Comment (gegaowp) [inline crates/sui-framework/sources/governance/sui_system.move:492]: are `storage_fund_balance` and `computation_charge` updated after each txn execution? are they reset around end of epochs?

File: crates/sui-framework/sources/governance/staking_pool.move [hunk 2]
```diff
@@ -415,6 +421,12 @@ module sui::staking_pool {
 
     public fun delegation_token_amount(delegation: &Delegation): u64 { balance::value(&delegation.pool_tokens) }
 
+    public fun pool_token_exchange_rate(pool: &StakingPool): PoolTokenExchangeRate {
```
Comment (gegaowp) [inline crates/sui-framework/sources/governance/staking_pool.move:424]: so to get the latest exchange rate, the "client" wants to make a move call of this function?

File: crates/sui-framework/sources/governance/sui_system.move [hunk 3]
```diff
@@ -21,6 +21,7 @@ module sui::sui_system {
     use sui::epoch_time_lock::EpochTimeLock;
     use sui::epoch_time_lock;
     use sui::pay;
+    use sui::event;
```
Comment (gegaowp) [inline crates/sui-framework/sources/governance/sui_system.move:24]: noob q, how / when are functions in `sui_system` called, by a blessed validator or sth else?

**Review by gegaowp — APPROVED** [review]
lgtm! looks like `tallying_rule_global_score` and `epoch` advance are not implemented yet, assuming they will be done in a follow-up PR.

File: crates/sui-framework/sources/governance/staking_pool.move (hunk 2 above)
Comment (666lcz) [reply crates/sui-framework/sources/governance/staking_pool.move:424]: It is also available as an event emitted by `emit_validator_epoch_events`, which is called by `advance_epoch`

**Review by 666lcz — APPROVED** [review]
Thanks for adding these! Confirmed that these suffice all the p0 items for the Explorer metrics as well.

**Comment by emmazzz** [comment]
> looks like tallying_rule_global_score is not implemented yet, assuming it will be done in a follow-up PR.
Yes, will be in this PR #7310

File: crates/sui-framework/sources/governance/sui_system.move (hunk 3 above)
Comment (emmazzz) [reply crates/sui-framework/sources/governance/sui_system.move:24]: Depends on which function you are asking about. Delegation related ones are callable by anybody. Validator management/self-staking ones are callable by validators. `advance_epoch` is special and only called during reconfig from Rust.

File: crates/sui-framework/sources/governance/staking_pool.move (hunk 2 above)
Comment (emmazzz) [reply crates/sui-framework/sources/governance/staking_pool.move:424]: > to get the latest exchange rate

The exchange rate is updated once per epoch during epoch change so the rate in the event is the latest

File: crates/sui-framework/sources/governance/sui_system.move (hunk 1 above)
Comment (emmazzz) [reply crates/sui-framework/sources/governance/sui_system.move:492]: > are storage_fund_balance and computation_charge updated after each txn execution?

No they come in once every epoch when this function `advance_epoch` is called.

**Pull request merged**